crates/mcp-client/src/service.rs

```diff
@@ -0,0 +1,111 @@
+
+use crate::transport::{Error as TransportError, Transport};
+use mcp_core::protocol::{JsonRpcMessage, JsonRpcResponse};
+
+#[derive(Debug, thiserror::Error)]
+pub enum ServiceError {
+    #[error("Transport error: {0}")]
+    Transport(#[from] TransportError),
+
+    #[error("Serialization error: {0}")]
+    Serialization(#[from] serde_json::Error),
+
+    #[error("Request timed out")]
+    Timeout(#[from] tower::timeout::error::Elapsed),
+
+    #[error("Other error: {0}")]
+    Other(String),
+
+    #[error("Unexpected server response")]
+    UnexpectedResponse,
+}
+
+/// A Tower `Service` implementation that uses a `Transport` to send/receive JsonRpcMessages and JsonRpcMessages.
+pub struct TransportService<T: Transport> {
+    transport: Arc<Mutex<T>>,
+    initialized: AtomicBool,
+}
+
+impl<T: Transport> TransportService<T> {
+    pub fn new(transport: T) -> Self {
+        Self {
+            transport: Arc::new(Mutex::new(transport)),
+            initialized: AtomicBool::new(false),
+        }
+    }
+
+    /// Provides a clone of the transport handle for external access (e.g., for sending notifications).
+    pub fn get_transport_handle(&self) -> Arc<Mutex<T>> {
+        Arc::clone(&self.transport)
+    }
+}
+
+impl<T: Transport> Service<JsonRpcMessage> for TransportService<T> {
+    type Response = JsonRpcMessage;
+    type Error = ServiceError;
+    type Future = Pin<Box<dyn Future<Output = Result<JsonRpcMessage, ServiceError>> + Send>>;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // Always ready. We do on-demand initialization in call().
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, message: JsonRpcMessage) -> Self::Future {
+        let transport = Arc::clone(&self.transport);
+        let started = self.initialized.load(Ordering::SeqCst);
+
+        Box::pin(async move {
+            let transport = transport.lock().await;
+
+            // Initialize (start) transport on the first call.
+            if !started {
+                transport.start().await?;
+            }
+
+            match message {
+                JsonRpcMessage::Notification(notification) => {
+                    // Serialize notification
+                    let msg = serde_json::to_string(&notification)?;
+                    transport.send(msg).await?;
+                    // For notifications, the protocol does not require a response
+                    // So we return an empty response here and this is not checked upstream
+                    let response: JsonRpcMessage = JsonRpcMessage::Response(JsonRpcResponse {
+                        jsonrpc: "2.0".to_string(),
+                        id: None,
+                        result: None,
+                        error: None,
+                    });
+
+                    Ok(response)
+                }
+                JsonRpcMessage::Request(request) => {
+                    // Serialize request & wait for response
```

Comment:
I don't see how this version works, but i might be missing something? 

It definitely works if everything is used in serial, a request gets sent in, the next message to receive is the response to that request, this exits happily.

But if we send multiple requests in parallel, i don't see that there's any guarantee they come back in order?

I think we need something like the previous Session implementation where each outbound request goes with an attached response_tx, so that this can await the response_rx.

Reply:
you're right! i rewrote this here: https://github.com/block/goose/pull/505